src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -424,9 +424,28 @@ private ThreadPoolWorkQueueThreadLocals CreateThreadLocals()
             return ThreadPoolWorkQueueThreadLocals.threadLocals = new ThreadPoolWorkQueueThreadLocals(this);
         }
 
+        internal void EnsureThreadPoolActive()
+        {
+            // Used for queuing.
+
+            // If we have not yet requested a thread, then request a new thread so the ThreadPool is active.
+            int count = numOutstandingThreadRequests;
+            while (count < 1)
```

Comment:
I believe this speculative check is insufficient to guarantee that the thread pool would be active on an enqueue.

I'm already looking into this problem and I think I have a decent approach that would be correct along with compensating for the extra overhead of a non-speculative check here. I'm currently looking at doing that along with my changes to switch the coreclr runtime to use the portable thread pool (WIP changes in my branch `ThreadPool` at the moment), as this type of change would involve a fair bit of testing that I would need to do with those changes anyway. If for some reason the portable thread pool change cannot be merged for 5.0 then there are several portions of that PR that I would plan to submit separately including a fix for this issue. Let me try a fix for it and see where it goes. If you have a scenario in mind where this type of change would be impactful I would greatly appreciate it if you could help to verify those.

Reply:
Edited above

Reply:
👍

Reply:
> If for some reason the portable thread pool change cannot be merged for 5.0 then there are several portions of that branch that I would plan to submit separately including a fix for this issue. Let me try a fix for it and see where it goes. If you have a scenario in mind where this type of change would be impactful I would greatly appreciate it if you could help to verify those, though that branch is not ready for verification yet. I'll ping on on the PR once I have it up.

No negative thoughts, we'll get it in for 5.0 😄

Reply:
@kouvel where's your branch?

Reply:
It's this one: https://github.com/kouvel/runtime/tree/ThreadPoolRequestFewerThreads
I will likely update the various commits with force-push so the commits may change over time (to make it easier to review) until I put up a PR.

Reply:
Eh that's the wrong one

Reply:
Updated inline above

Reply:
It doesn't have a fix for this issue yet, I have a prototype that I'll be adding to that branch

Reply:
Awesome! I'll be stalking these changes 😄 .